### ISSUE TITLE
Portable binaries for Mac OS and Linux

### ISSUE THREAD

**Issue by vanka78bg** [libsdl-org/SDL#4249]
Hello,

I am trying to build portable binaries for Linux and Mac OS from source for use with C# and [SDL2#](https://github.com/flibitijibibo/SDL2-CS) .  I am using the following shell commands, according to the documentation:

```
./configure --prefix=<install-path>
make
make install
```

The above works and generates the necessary *.so and *.dylib libraries. My problem occurs when trying to copy my application to another machine, without these libraries pre-installed. I have tried to include the binaries with my application, so no extra installation steps are required. The SDL2# library loads the corresponding libSDL2.so or libSDL2.dylib successfully, but fails to load libSDL2_image.so/libSDL2_image.dylib.

After some digging I have found that the SDL2_image library simply cannot find the core SDL2 library, even though both libraries are at the same location under the application folder. Apparently SDL2_image links against SDL2 using an absolute path. If SDL2 is not installed on the exact same location on the target machine, SDL2_image cannot find it.

I have tried to use `install_name_tool` to correct this issue, without any success so far:

```
install_name_tool -change <old-install-path> @executable_path/libSDL2.dylib libSDL2_image.dylib
```

The above command is for Mac OS. Unfortunately this does not work. What am I doing wrong?

Any help would be greatly appreciated.

**Comment by sezero** [comment]
I'd first do `install_name_tool -id @executable_path/libSDL2.dylib libSDL2.dylib`,
then build SDL_image against this libSDL2.dylib and then do:
`install_name_tool -id @executable_path/libSDL2_image.dylib libSDL2_image.dylib`,
and then build my apps against these install-name-corrected dylibs.

**Comment by vanka78bg** [comment]
Thank you for the quick reply! I'll try this as soon as I have the chance. I was sure I was missing something obvious. :-)

**Comment by vanka78bg** [comment]
I have tried the above, but now I have faced another problem. Since I am modifying the id of the libSDL2.dylib in-place, before building libSDL2_image.dylib, I would prefer to install libSDL2.dylib to another location (different from /usr/local/lib), so I do not mess up something (this is a part of an automated build). What I do is to specify a custom prefix in the configure script when building SDL:

```
cd sdl
./configure --prefix=$PWD/../dist
make
make install
cd ../dist/lib
install_name_tool -id @executable_path/libSDL2.dylib libSDL2.dylib
```

Then I try to build SDL_image like this:

```
cd ../sdl_image
./configure --prefix=$PWD/../dist
make
make install
```

Now, the configure fails with an error saying it cannot find SDL:

```
checking for SDL - version >= 2.0.8... no
*** Could not run SDL test program, checking why...
*** The test program compiled, but did not run. This usually means
*** that the run-time linker is not finding SDL or finding the wrong
*** version of SDL. If it is not finding SDL, you'll need to set your
*** LD_LIBRARY_PATH environment variable, or edit /etc/ld.so.conf to point
*** to the installed location  Also, make sure you have run ldconfig if that
*** is required on your system
***
*** If you have an old version installed, it is best to remove it, although
*** you may also be able to get things to work by modifying LD_LIBRARY_PATH
configure: error: *** SDL version 2.0.8 not found!
```

I have tried to modify `LD_LIBRARY_PATH`, appending my custom path to it, as the error suggested, to no avail - the same error occurs every time. Is there any way to specify a custom location for SDL when building SDL_Image?

**Comment by sezero** [comment]
Sorry, didn't read most of your message, but: __if__ I remember correctly,
for mac `LD_LIBRARY_PATH` should be replaced with `DYLD_LIBRARY_PATH`
(can't be sure though.)

**Comment by vanka78bg** [comment]
You are correct about the name, however appending the path to `DYLD_LIBRARY_PATH` did not fix the problem as well. I am curious how SDL_image CI build handles this, since it is not currently migrated to GitHub actions?

**Comment by vanka78bg** [comment]
OK, I think I have figured this out at last. Instead of using `@executable_path`, I should use `@loader_path`. Here is the list of commands that made it work:

```
install_name_tool -id @loader_path/libSDL2.dylib libSDL2.dylib
install_name_tool -id @loader_path/libSDL2_image.dylib libSDL2_image.dylib
install_name_tool -change /usr/local/lib/libSDL2-2.0.0.dylib @loader_path/libSDL2.dylib libSDL2_image.dylib
```

Perhaps, changing just the id of libSDL2.dylib and building libSDL2_image.dylib against that modified library would also work, but as I have said, I am hesitant to modify libraries directly in /usr/local/lib, so I prefer to work with the local copy in my project instead.

I am closing this issue, since I believe I have resolved it for my scenario.
